staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -302,27 +305,56 @@ func translateTopologyFromCSIToInTree(pv *v1.PersistentVolume, topologyKey strin
 				}
 			}
 		} else {
-			// nothing is in the label, we try our best to add the CSI topology zone in
+			// nothing is in the zone label, we try our best to add the zone topology
 			if len(csiTopologyZoneValues) > 0 {
 				err := addTopology(pv, zoneLabel, csiTopologyZoneValues)
 				if err != nil {
 					return fmt.Errorf("Failed to add Kubernetes topology zone to PV NodeAffinity. %v", err)
 				}
 			}
 		}
+	}
+
+	// Check if Kubernetes Region Topology already exist
+	topologyRegionValue := getTopologyValues(pv, regionLabel)
+	if len(topologyRegionValue) == 0 {
 		if regionOK {
 			err := addTopology(pv, regionLabel, []string{regionLabelVal})
 			if err != nil {
 				return fmt.Errorf("Failed to add Kubernetes topology region to PV NodeAffinity. %v", err)
 			}
+		} else {
+			// nothing is in the region label, we try our best to add the region topology
+			if len(csiTopologyZoneValues) > 0 && regionParser != nil {
+				regionVal, err := regionParser(csiTopologyZoneValues)
+				if err != nil {
+					return fmt.Errorf("Failed to parse zones value(%v) to region label %v", csiTopologyZoneValues, err)
+				}
+				err = addTopology(pv, regionLabel, []string{regionVal})
+				if err != nil {
+					return fmt.Errorf("Failed to add Kubernetes topology region to PV NodeAffinity. %v", err)
+				}
+			}
 		}
 	}
 
 	// 3. Add Kubernetes Topology labels, if it already exists, we trust it
-	if !zoneOK {
-		pv.Labels[zoneLabel] = csiTopologyZoneValStr
+	if len(csiTopologyZoneValues) > 0 {
+		if pv.Labels == nil {
+			pv.Labels = make(map[string]string)
+		}
+		if !zoneOK {
+			csiTopologyZoneValStr := strings.Join(csiTopologyZoneValues, labelMultiZoneDelimiter)
+			pv.Labels[zoneLabel] = csiTopologyZoneValStr
+		}
+		if !regionOK && regionParser != nil {
+			regionVal, err := regionParser(csiTopologyZoneValues)
```

Comment:
It may not always be the case that you can derive region from zone value. It may be better to just pass the whole original PV object in, and potentially let the plugin extract region from their own topology label.

Reply:
Done.